src/convenience/session.ts

```diff
@@ -173,41 +216,55 @@ export interface SessionOptions<S> {
-    const getSessionKey = options.getSessionKey ?? defaultGetSessionKey;
-    const storage = options.storage ??
-        (debug(
-            "Storing session data in memory, all data will be lost when the bot restarts.",
-        ),
-            new MemorySessionStorage());
+    return options.type === "multi"
+        ? strictMultiSession(options)
+        : strictSingleSession(options);
+}
+
+function strictSingleSession<S, C extends Context>(
+    options: SessionOptions<S>,
+): MiddlewareFn<C & SessionFlavor<S>> {
+    const { initial, storage, getSessionKey, custom } = fillDefaults(options);
     return async (ctx, next) => {
+        const propSession = new PropertySession<SessionFlavor<S>, "session">(
+            storage,
+            ctx,
+            "session",
+            initial,
+        );
         const key = await getSessionKey(ctx);
-        let value = key === undefined
-            ? undefined
-            : (await storage.read(key)) ?? options.initial?.();
-        Object.defineProperty(ctx, "session", {
-            enumerable: true,
-            get() {
-                if (key === undefined) {
-                    const msg = undef("access", getSessionKey);
-                    throw new Error(msg);
-                }
-                return value;
-            },
-            set(v) {
-                if (key === undefined) {
-                    const msg = undef("assign", getSessionKey);
-                    throw new Error(msg);
-                }
-                value = v;
-            },
-        });
+        await propSession.init(key, { custom, lazy: false });
         await next(); // no catch: do not write back if middleware throws
-        if (key !== undefined) {
-            if (value == null) await storage.delete(key);
-            else await storage.write(key, value);
-        }
+        await propSession.finish();
+    };
+}
+function strictMultiSession<S, C extends Context>(
+    options: MultiSessionOptions<S>,
+): MiddlewareFn<C & SessionFlavor<S>> {
+    const props = Object.keys(options).filter((k) => k !== "type");
+    const defaults = Object.fromEntries(
+        props.map((prop) => [prop, fillDefaults(options[prop])]),
+    );
+    return async (ctx, next) => {
+        ctx.session = {} as S;
+        const propSessions = await Promise.all(props.map(async (prop) => {
+            const { initial, storage, getSessionKey, custom } = defaults[prop];
+            const s = new PropertySession(
+                // @ts-ignore cannot express that the storage works for a concrete prop
+                storage,
+                ctx.session,
+                prop,
+                initial,
+            );
+            const key = await getSessionKey(ctx);
+            await s.init(key, { custom, lazy: false });
+            return s;
+        }));
+        await next(); // no catch: do not write back if middleware throws
+        if (ctx.session == null) propSessions.forEach((s) => s.delete());
```

Comment:
shouldn't we add await here?

Reply:
I see, it's synchronous

Reply:
It's like doing `ctx.session.foo = undefined` manually if people do `ctx.session`.